core/backend/src/IModelDb.ts

```diff
@@ -809,22 +836,60 @@ export abstract class IModelDb extends IModel {
    * @see querySchemaVersion
    */
   public async importSchemas(schemaFileNames: LocalFileName[], options?: SchemaImportOptions): Promise<void> {
-    if (this.nativeDb.getITwinId() !== Guid.empty) // if this iModel is associated with an iTwin, importing schema requires the schema lock
-      await this.acquireSchemaLock();
+    if (schemaFileNames.length === 0)
+      return;
 
     const maybeCustomNativeContext = options?.ecSchemaXmlContext?.nativeContext;
-    const nativeImportOptions: IModelJsNative.SchemaImportOptions = {
-      schemaLockHeld: true,
-      ecSchemaXmlContext: maybeCustomNativeContext,
-    };
+    if (this.schemaSyncEnabled()) {
+      const store = this._schemaSyncAccess;
+      if (!store) {
+        throw new IModelError(DbResult.BE_SQLITE_ERROR, "Schema sync db store cloud access is not set.");
+      }
+      await store.withLockedDb("initialize schema synchronization", async () => {
+        const schemaSyncDbUri = store.getUri();
 
-    const stat = this.nativeDb.importSchemas(schemaFileNames, nativeImportOptions);
-    if (DbResult.BE_SQLITE_OK !== stat) {
-      throw new IModelError(stat, "Error importing schema");
-    }
+        this.saveChanges();
+        let stat = this.nativeDb.importSchemas(schemaFileNames, { schemaLockHeld: false, ecSchemaXmlContext: maybeCustomNativeContext, schemaSyncDbUri });
+        if (DbResult.BE_SQLITE_ERROR_SchemaLockFailed === stat) {
+          this.abandonChanges();
+          if (this.nativeDb.getITwinId() !== Guid.empty)
+            await this.acquireSchemaLock();
+          stat = this.nativeDb.importSchemas(schemaFileNames, { schemaLockHeld: true, ecSchemaXmlContext: maybeCustomNativeContext, schemaSyncDbUri });
+        }
+        if (DbResult.BE_SQLITE_OK !== stat) {
+          throw new IModelError(stat, "Error importing schema");
+        }
+      }, OpenMode.Readonly);
+    } else {
+      const nativeImportOptions: IModelJsNative.SchemaImportOptions = {
+        schemaLockHeld: true,
+        ecSchemaXmlContext: maybeCustomNativeContext,
+      };
 
+      if (this.nativeDb.getITwinId() !== Guid.empty) // if this iModel is associated with an iTwin, importing schema requires the schema lock
+        await this.acquireSchemaLock();
+
+      const stat = this.nativeDb.importSchemas(schemaFileNames, nativeImportOptions);
+      if (DbResult.BE_SQLITE_OK !== stat) {
+        throw new IModelError(stat, "Error importing schema");
+      }
+    }
     this.clearCaches();
   }
+  /** @internal */
+  public synchronizationSchemas() {
+    if (this.schemaSyncEnabled()) {
+      const store = this._schemaSyncAccess;
+      if (!store) {
+        throw new IModelError(DbResult.BE_SQLITE_ERROR, "Schema sync db cloud access is not set.");
+      }
+      store.synchronizeWithCloud();
+      store.openForRead();
+      this.nativeDb.schemaSyncPull(store.getUri());
+      // we need to make sure if there was any change at all that was pulled rather then clear cache every time.
```

Comment:
probably should close the syncDb here.

Reply:
I have remove temp var where they were not necessary and renamed it to syncDb where it was needed.